core/src/components/datetime/datetime.tsx

```diff
@@ -1397,7 +1403,8 @@ export class Datetime implements ComponentInterface {
    * should just be the default segment.
    */
   private renderTime(mode: Mode) {
-    const use24Hour = is24Hour(this.locale);
+    const { timeFormat } = this;
+    const use24Hour = timeFormat != null ? timeFormat === '24h' : is24Hour(this.locale);
```

Comment:
We are going to need to revamp the `is24Hour` function a bit since you can also encode the hour cycle inside of the locale string. (See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle#adding_an_hour_cycle_via_the_locale_string)

My current thinking is pass `this.hourCycle` and `this.locale` into `is24Hour`. The internals of the function would look something like this:

```typescript
export const is24Hour = (locale: string, hourCycle?: 'h23' | 'h12') => {
  /**
   * If developer has explicitly enabled h23 time
   * then return early and do not look at the system default.
   */
  if (hourCycle !== undefined) {
    return hourCycle === 'h23';
  }

  /**
   * If hourCycle was not specified, check the locale
   * that is set on the user's device. We first check the
   * Intl.DateTimeFormat hourCycle option as developers can encode this
   * option into the locale string. Example: `en-US-u-hc-h23`
   */
  const date = new Date('5/18/2021 00:00');
  const formatted = new Intl.DateTimeFormat(locale, { hour: 'numeric' });
  const options = formatted.resolvedOptions();
  if (options.hourCycle !== undefined) {
    return options.hourCycle === 'h23';
  }
  
  /**
   * If hourCycle is not specified (either through lack
   * of browser support or locale information) then fall
   * back to this slower hourCycle check.
   */
  const parts = formatted.formatToParts(date);
  const hour = parts.find(p => p.type === 'hour');

  if (!hour) {
    throw new Error('Hour value not found from DateTimeFormat');
  }

  return hour.value === '00';
}
```

The support for `options.hourCycle` is somewhat new, but as time goes on we would be able to rely more and more on this instead of our fallback hack.

This would need some more revamping if we are to support `h11` and `h24` formats, but I think sticking with `h12` and `h23` is fine for now.

Also please add tests to the `helpers.spec.ts` file for this.

I know things get a bit tricky when you are dealing with different hour formats, so let me know if you would like any assistance on this change.

Reply:
I get this error by my idea when trying to adjust the function: 

**TS2339: Property 'hourCycle' does not exist on type 'ResolvedDateTimeFormatOptions'**

and found this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/resolvedOptions

Are you sure this is available via the `formatted.resolvedOptions()`?

Reply:
Or do we have to use this there instead?

```
let locale = new Intl.Locale("fr-FR-u-hc-h23");
console.log(locale.hourCycle);
```

Reply:
MDN is out of date it looks like: https://caniuse.com/mdn-javascript_builtins_intl_datetimeformat_datetimeformat_options_parameter_options_hourcycle_parameter

Reply:
Trying this

`let locale = new Intl.Locale("fr-FR-u-hc-h23");`

also says **TS2339: Property 'Locale' does not exist on type 'typeof Intl'**

Reply:
TypeScript doesn't recognize it, but it should: https://github.com/microsoft/TypeScript/issues/34399 

In that case, we could extend `ResolvedDateTimeFormatOptions` and cast the result as the custom interface:

```typescript
/**
 * Typescript 4.x does not recognize hourCycle as a valid option.
 *  See https://github.com/microsoft/TypeScript/issues/34399.
 */
interface DatetimeFormatOptions extends ResolvedDateTimeFormatOptions {
  hourCycle?: 'h11' | 'h12' | 'h23' | 'h24'
}

const options = formatted.resolvedOptions() as DatetimeFormatOptions;
```

Reply:
Okay i will add this

Reply:
We use the check here to determinate if we should use AM/PM or not and the hour values right?
Shouldn't we check then:

```
if (options.hourCycle !== undefined) {
    return options.hourCycle === 'h23' || options.hourCycle === 'h24';
  }
```

as the `h24` also should not show AM/PM currently?

Reply:
I think for the purposes of this PR we can support `h12` and `h23` for now. I am not opposed to supporting `h11` and `h24` but it seems like `h12` and `h23` are the more common formats that Ionic developers are looking to use.

Reply:
If you want to add support for `h11` and `h24` definitely feel free to add it, but that's not a requirement for this feature to be merged.

Reply:
Yes, i don't mean to add `h11` and `h24` in general, i meant if someone would pass a local like `fr-FR-u-hc-h24` the above implementation would show AM/PM

Reply:
This would be easily changed by added this to the check here:

```ts
/**
 * If hourCycle was not specified, check the locale
 * that is set on the user's device. We first check the
 * Intl.DateTimeFormat hourCycle option as developers can encode this
 * option into the locale string. Example: `en-US-u-hc-h23`
 */
const formatted = new Intl.DateTimeFormat(locale, { hour: 'numeric' });
const options = formatted.resolvedOptions() as DatetimeFormatOptions;
if (options.hourCycle !== undefined) {
  return options.hourCycle === 'h23' || options.hourCycle === 'h24';
}
```